menoh/tensorrt/Inference.cpp

```diff
@@ -89,15 +96,143 @@ namespace menoh_impl {
                        totalTime / TIMING_ITERATIONS);
             }
         } gProfiler;
-#endif
+#endif // MENOH_ENABLE_TENSORRT_PROFILER
         static Logger gLogger;
 
+#ifdef MENOH_ENABLE_TENSORRT_MODEL_CACHING
+        std::string Inference::calc_model_hash(
+          std::unordered_map<std::string, array> const& input_table,
+          std::unordered_map<std::string, array> const& output_table,
+          menoh_impl::model_data const& model_data, config const& config) {
+            hasher hasher;
+            auto add_variable_table =
+              [](menoh_impl::hasher& h,
+                 std::unordered_map<std::string, array> const& table) {
+                  std::vector<std::pair<std::string, array>> variables(
+                    table.begin(), table.end());
+                  std::sort(variables.begin(), variables.end(),
+                            [](auto const& a, auto const& b) {
+                                return a.first < b.first;
+                            });
+                  for(auto const& p : variables) {
+                      add_container(h, p.first);
+                      h.add(static_cast<const std::uint8_t*>(
+                              static_cast<const void*>(p.second.dims().data())),
+                            p.second.dims().size());
+                      // Do not process values in p.second because it is a
+                      // placeholder
+                  }
+              };
+            add_variable_table(hasher, input_table);  // [input_table]
+            add_variable_table(hasher, output_table); // [output_table]
+            auto add_str_vec = [](menoh_impl::hasher& h,
+                                  std::vector<std::string> const& sv) {
+                std::for_each(sv.begin(), sv.end(),
+                              [&h](auto const& s) { add_container(h, s); });
+            };
+            auto add_attr = [](menoh_impl::hasher& h, auto const& a) {
+                int* i;
+                float* f;
+                std::vector<int>* is;
+                std::vector<float>* fs;
+                if((i = const_cast<int*>(get_if<int>(&a)))) {
+                    add_container(h, "int" + std::to_string(*i));
+                } else if((f = const_cast<float*>(get_if<float>(&a)))) {
+                    add_container(h, "float" + std::to_string(*f));
+                } else if((is = const_cast<std::vector<int>*>(
+                             get_if<std::vector<int>>(&a)))) {
+                    add_c_str(h, "ints");
+                    std::for_each(is->begin(), is->end(), [&h](auto i) {
+                        add_container(h, std::to_string(i));
+                    });
+                } else if((fs = const_cast<std::vector<float>*>(
+                             get_if<std::vector<float>>(&a)))) {
```

Comment:
Isn't it possible to declare those pointer variables as `const` and remove those `const_cast`s?

Reply:
OK. I'll try.